{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using ezancestry as a Python library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load config variables\n",
    "from ezancestry.config import aisnps_directory as _aisnps_directory\n",
    "from ezancestry.config import aisnps_set as _aisnps_set\n",
    "from ezancestry.config import algorithm as _algorithm\n",
    "from ezancestry.config import k as _k\n",
    "from ezancestry.config import models_directory as _models_directory\n",
    "from ezancestry.config import n_components as _n_components\n",
    "from ezancestry.config import population_level as _population_level\n",
    "from ezancestry.config import samples_directory as _samples_directory\n",
    "from ezancestry.config import thousand_genomes_directory as _thousand_genomes_directory\n",
    "\n",
    "# load functions\n",
    "from ezancestry.aisnps import extract_aisnps\n",
    "from ezancestry.dimred import dimensionality_reduction\n",
    "from ezancestry.evaluate import export_performance\n",
    "from ezancestry.fetch import download_thousand_genomes\n",
    "from ezancestry.model import predict_ancestry, train\n",
    "from ezancestry.process import (encode_genotypes, get_1kg_labels,\n",
    "                                process_user_input, vcf2df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pull aisnps from 1kG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/jacksonc08/.ezancestry/data/aisnps/kidd.aisnp.txt\n"
     ]
    }
   ],
   "source": [
    "# # kidd\n",
    "aisnps_file = Path(_aisnps_directory).joinpath(\"kidd.aisnp.txt\")\n",
    "print(aisnps_file)\n",
    "#extract_aisnps(_thousand_genomes_directory, aisnps_file, aisnps_set=\"kidd\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Seldin#\n",
    "aisnps_file = Path(_aisnps_directory).joinpath(\"Seldin.aisnp.txt\")\n",
    "#extract_aisnps(_thousand_genomes_directory, aisnps_file, aisnps_set=\"Seldin\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull the 1000 Genomes Project samples\n",
    "dfsamples = get_1kg_labels(_samples_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>superpopulation</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        population superpopulation  gender\n",
       "sample                                    \n",
       "HG00096        GBR             EUR    male\n",
       "HG00097        GBR             EUR  female\n",
       "HG00099        GBR             EUR  female\n",
       "HG00100        GBR             EUR  female\n",
       "HG00101        GBR             EUR    male"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfsamples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3737576</th>\n",
       "      <th>rs7554936</th>\n",
       "      <th>rs2814778</th>\n",
       "      <th>rs4918664</th>\n",
       "      <th>rs174570</th>\n",
       "      <th>rs1079597</th>\n",
       "      <th>rs2238151</th>\n",
       "      <th>rs671</th>\n",
       "      <th>rs7997709</th>\n",
       "      <th>rs1572018</th>\n",
       "      <th>...</th>\n",
       "      <th>rs3823159</th>\n",
       "      <th>rs917115</th>\n",
       "      <th>rs1462906</th>\n",
       "      <th>rs6990312</th>\n",
       "      <th>rs2196051</th>\n",
       "      <th>rs1871534</th>\n",
       "      <th>rs3814134</th>\n",
       "      <th>population</th>\n",
       "      <th>superpopulation</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        rs3737576 rs7554936 rs2814778 rs4918664 rs174570 rs1079597 rs2238151  \\\n",
       "HG00096        TT        CT        TT        AG       CC        CC        TT   \n",
       "HG00097        TT        TT        TT        AA       CC        CC        CC   \n",
       "HG00099        TT        CT        TT        AA       CC        CC        CT   \n",
       "HG00100        TT        TT        TT        AG       CT        CC        TT   \n",
       "HG00101        TT        CT        TT        AA       CC        CC        TT   \n",
       "\n",
       "        rs671 rs7997709 rs1572018  ... rs3823159 rs917115 rs1462906 rs6990312  \\\n",
       "HG00096    GG        TT        CC  ...        AA       TT        CC        GG   \n",
       "HG00097    GG        CT        CC  ...        AA       TT        CC        GT   \n",
       "HG00099    GG        TT        CC  ...        AA       TT        CC        GG   \n",
       "HG00100    GG        TT        CC  ...        AA       TT        CC        GT   \n",
       "HG00101    GG        TT        CC  ...        AA       CT        CC        GG   \n",
       "\n",
       "        rs2196051 rs1871534 rs3814134 population superpopulation  gender  \n",
       "HG00096        AA        CC        AA        GBR             EUR    male  \n",
       "HG00097        AG        CC        AA        GBR             EUR  female  \n",
       "HG00099        AG        CC        AA        GBR             EUR  female  \n",
       "HG00100        AG        CC        AA        GBR             EUR  female  \n",
       "HG00101        AA        CC        AA        GBR             EUR    male  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_fname = Path(_aisnps_directory).joinpath(\"kidd.aisnp.1kg.vcf\")\n",
    "df_kidd = vcf2df(vcf_fname, dfsamples)\n",
    "\n",
    "df_kidd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n",
      "/Users/jacksonc08/ezancestry/ezancestry/process.py:57: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[variant.ID] = [\n"
     ]
    }
   ],
   "source": [
    "vcf_fname = Path(_aisnps_directory).joinpath(\"Seldin.aisnp.1kG.vcf\")\n",
    "df_seldin = vcf2df(vcf_fname, dfsamples)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### could start here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_kidd, test_kidd, y_train_kidd, y_test_kidd = train_test_split(\n",
    "    df_kidd,\n",
    "    df_kidd[\"superpopulation\"],\n",
    "    test_size=0.2,\n",
    "    stratify=df_kidd[\"superpopulation\"],\n",
    "    random_state=42,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### one-hot encode snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The user could have missing snps\n",
    "df_user = df_kidd[df_kidd.columns[0:43]].copy()\n",
    "\n",
    "# The user could have extra snps\n",
    "df_user[\"extra_snp\"] = \"TT\"\n",
    "\n",
    "# The user could have genotypes that weren't in the original encoder\n",
    "df_user.loc[\"HG00096\", \"rs3737576\"] = \"blah\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3737576</th>\n",
       "      <th>rs7554936</th>\n",
       "      <th>rs2814778</th>\n",
       "      <th>rs4918664</th>\n",
       "      <th>rs174570</th>\n",
       "      <th>rs1079597</th>\n",
       "      <th>rs2238151</th>\n",
       "      <th>rs671</th>\n",
       "      <th>rs7997709</th>\n",
       "      <th>rs1572018</th>\n",
       "      <th>...</th>\n",
       "      <th>rs260690</th>\n",
       "      <th>rs6754311</th>\n",
       "      <th>rs10497191</th>\n",
       "      <th>rs310644</th>\n",
       "      <th>rs2024566</th>\n",
       "      <th>rs12498138</th>\n",
       "      <th>rs4833103</th>\n",
       "      <th>rs1229984</th>\n",
       "      <th>rs3811801</th>\n",
       "      <th>extra_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>blah</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AC</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        rs3737576 rs7554936 rs2814778 rs4918664 rs174570 rs1079597 rs2238151  \\\n",
       "HG00096      blah        CT        TT        AG       CC        CC        TT   \n",
       "HG00097        TT        TT        TT        AA       CC        CC        CC   \n",
       "HG00099        TT        CT        TT        AA       CC        CC        CT   \n",
       "HG00100        TT        TT        TT        AG       CT        CC        TT   \n",
       "HG00101        TT        CT        TT        AA       CC        CC        TT   \n",
       "\n",
       "        rs671 rs7997709 rs1572018  ... rs260690 rs6754311 rs10497191 rs310644  \\\n",
       "HG00096    GG        TT        CC  ...       AC        TT         CC       CT   \n",
       "HG00097    GG        CT        CC  ...       CC        TT         CC       TT   \n",
       "HG00099    GG        TT        CC  ...       AA        TT         CC       TT   \n",
       "HG00100    GG        TT        CC  ...       AA        CT         CC       TT   \n",
       "HG00101    GG        TT        CC  ...       AA        CC         CC       TT   \n",
       "\n",
       "        rs2024566 rs12498138 rs4833103 rs1229984 rs3811801 extra_snp  \n",
       "HG00096        AA         GG        AC        CC        GG        TT  \n",
       "HG00097        AG         GG        AC        CC        GG        TT  \n",
       "HG00099        AG         GG        AC        CC        GG        TT  \n",
       "HG00100        AA         AG        AA        CC        GG        TT  \n",
       "HG00101        AG         AG        AC        CC        GG        TT  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_user.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:51.133 | INFO     | ezancestry.process:encode_genotypes:137 - Successfully loaded an encoder from /Users/jacksonc08/.ezancestry/data/models/one_hot_encoder.kidd.bin\n"
     ]
    }
   ],
   "source": [
    "ohe_user = encode_genotypes(df_user, aisnps_set=\"kidd\", overwrite_encoder=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3737576_CC</th>\n",
       "      <th>rs3737576_CT</th>\n",
       "      <th>rs3737576_TT</th>\n",
       "      <th>rs7554936_CC</th>\n",
       "      <th>rs7554936_CT</th>\n",
       "      <th>rs7554936_TT</th>\n",
       "      <th>rs2814778_CC</th>\n",
       "      <th>rs2814778_CT</th>\n",
       "      <th>rs2814778_TT</th>\n",
       "      <th>rs798443_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>rs4891825_GG</th>\n",
       "      <th>rs7251928_AA</th>\n",
       "      <th>rs7251928_AC</th>\n",
       "      <th>rs7251928_CC</th>\n",
       "      <th>rs310644_CC</th>\n",
       "      <th>rs310644_CT</th>\n",
       "      <th>rs310644_TT</th>\n",
       "      <th>rs2024566_AA</th>\n",
       "      <th>rs2024566_AG</th>\n",
       "      <th>rs2024566_GG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA21137</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA21141</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA21142</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA21143</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA21144</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2504 rows × 165 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rs3737576_CC  rs3737576_CT  rs3737576_TT  rs7554936_CC  rs7554936_CT  \\\n",
       "HG00096           0.0           0.0           0.0           0.0           1.0   \n",
       "HG00097           0.0           0.0           1.0           0.0           0.0   \n",
       "HG00099           0.0           0.0           1.0           0.0           1.0   \n",
       "HG00100           0.0           0.0           1.0           0.0           0.0   \n",
       "HG00101           0.0           0.0           1.0           0.0           1.0   \n",
       "...               ...           ...           ...           ...           ...   \n",
       "NA21137           0.0           0.0           1.0           0.0           0.0   \n",
       "NA21141           0.0           0.0           1.0           0.0           0.0   \n",
       "NA21142           0.0           0.0           1.0           0.0           0.0   \n",
       "NA21143           0.0           0.0           1.0           0.0           0.0   \n",
       "NA21144           0.0           0.0           1.0           0.0           1.0   \n",
       "\n",
       "         rs7554936_TT  rs2814778_CC  rs2814778_CT  rs2814778_TT  rs798443_AA  \\\n",
       "HG00096           0.0           0.0           0.0           1.0          1.0   \n",
       "HG00097           1.0           0.0           0.0           1.0          1.0   \n",
       "HG00099           0.0           0.0           0.0           1.0          0.0   \n",
       "HG00100           1.0           0.0           0.0           1.0          1.0   \n",
       "HG00101           0.0           0.0           0.0           1.0          1.0   \n",
       "...               ...           ...           ...           ...          ...   \n",
       "NA21137           1.0           0.0           0.0           1.0          1.0   \n",
       "NA21141           1.0           0.0           0.0           1.0          0.0   \n",
       "NA21142           1.0           0.0           0.0           1.0          0.0   \n",
       "NA21143           1.0           0.0           0.0           1.0          0.0   \n",
       "NA21144           0.0           0.0           0.0           1.0          1.0   \n",
       "\n",
       "         ...  rs4891825_GG  rs7251928_AA  rs7251928_AC  rs7251928_CC  \\\n",
       "HG00096  ...           0.0           0.0           1.0           0.0   \n",
       "HG00097  ...           0.0           1.0           0.0           0.0   \n",
       "HG00099  ...           0.0           1.0           0.0           0.0   \n",
       "HG00100  ...           0.0           1.0           0.0           0.0   \n",
       "HG00101  ...           0.0           0.0           1.0           0.0   \n",
       "...      ...           ...           ...           ...           ...   \n",
       "NA21137  ...           0.0           0.0           1.0           0.0   \n",
       "NA21141  ...           0.0           1.0           0.0           0.0   \n",
       "NA21142  ...           0.0           1.0           0.0           0.0   \n",
       "NA21143  ...           0.0           1.0           0.0           0.0   \n",
       "NA21144  ...           0.0           1.0           0.0           0.0   \n",
       "\n",
       "         rs310644_CC  rs310644_CT  rs310644_TT  rs2024566_AA  rs2024566_AG  \\\n",
       "HG00096          0.0          1.0          0.0           1.0           0.0   \n",
       "HG00097          0.0          0.0          1.0           0.0           1.0   \n",
       "HG00099          0.0          0.0          1.0           0.0           1.0   \n",
       "HG00100          0.0          0.0          1.0           1.0           0.0   \n",
       "HG00101          0.0          0.0          1.0           0.0           1.0   \n",
       "...              ...          ...          ...           ...           ...   \n",
       "NA21137          1.0          0.0          0.0           0.0           1.0   \n",
       "NA21141          0.0          0.0          1.0           1.0           0.0   \n",
       "NA21142          1.0          0.0          0.0           1.0           0.0   \n",
       "NA21143          1.0          0.0          0.0           1.0           0.0   \n",
       "NA21144          0.0          0.0          1.0           0.0           1.0   \n",
       "\n",
       "         rs2024566_GG  \n",
       "HG00096           0.0  \n",
       "HG00097           0.0  \n",
       "HG00099           0.0  \n",
       "HG00100           0.0  \n",
       "HG00101           0.0  \n",
       "...               ...  \n",
       "NA21137           0.0  \n",
       "NA21141           0.0  \n",
       "NA21142           0.0  \n",
       "NA21143           0.0  \n",
       "NA21144           0.0  \n",
       "\n",
       "[2504 rows x 165 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ohe_user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rs3737576_CC    0.0\n",
       "rs3737576_CT    0.0\n",
       "rs3737576_TT    0.0\n",
       "Name: HG00096, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure \"blah\" genotype didn't get encoded\n",
    "ohe_user.loc[\"HG00096\", [\"rs3737576_CC\", \"rs3737576_CT\", \"rs3737576_TT\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change to True to write new encoders\n",
    "OVERWRITE_ENCODER = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:51.219 | INFO     | ezancestry.process:encode_genotypes:137 - Successfully loaded an encoder from /Users/jacksonc08/.ezancestry/data/models/one_hot_encoder.kidd.bin\n",
      "2022-10-25 14:12:51.277 | INFO     | ezancestry.process:encode_genotypes:137 - Successfully loaded an encoder from /Users/jacksonc08/.ezancestry/data/models/one_hot_encoder.seldin.bin\n"
     ]
    }
   ],
   "source": [
    "# get an encoder for each snp set\n",
    "df_kidd_encoded = encode_genotypes(df_kidd, aisnps_set=\"kidd\", overwrite_encoder=OVERWRITE_ENCODER)\n",
    "df_seldin_encoded = encode_genotypes(df_seldin, aisnps_set=\"Seldin\", overwrite_encoder=OVERWRITE_ENCODER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3737576_CC</th>\n",
       "      <th>rs3737576_CT</th>\n",
       "      <th>rs3737576_TT</th>\n",
       "      <th>rs7554936_CC</th>\n",
       "      <th>rs7554936_CT</th>\n",
       "      <th>rs7554936_TT</th>\n",
       "      <th>rs2814778_CC</th>\n",
       "      <th>rs2814778_CT</th>\n",
       "      <th>rs2814778_TT</th>\n",
       "      <th>rs798443_AA</th>\n",
       "      <th>...</th>\n",
       "      <th>rs4891825_GG</th>\n",
       "      <th>rs7251928_AA</th>\n",
       "      <th>rs7251928_AC</th>\n",
       "      <th>rs7251928_CC</th>\n",
       "      <th>rs310644_CC</th>\n",
       "      <th>rs310644_CT</th>\n",
       "      <th>rs310644_TT</th>\n",
       "      <th>rs2024566_AA</th>\n",
       "      <th>rs2024566_AG</th>\n",
       "      <th>rs2024566_GG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00100</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00101</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 165 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rs3737576_CC  rs3737576_CT  rs3737576_TT  rs7554936_CC  rs7554936_CT  \\\n",
       "HG00096           0.0           0.0           1.0           0.0           1.0   \n",
       "HG00097           0.0           0.0           1.0           0.0           0.0   \n",
       "HG00099           0.0           0.0           1.0           0.0           1.0   \n",
       "HG00100           0.0           0.0           1.0           0.0           0.0   \n",
       "HG00101           0.0           0.0           1.0           0.0           1.0   \n",
       "\n",
       "         rs7554936_TT  rs2814778_CC  rs2814778_CT  rs2814778_TT  rs798443_AA  \\\n",
       "HG00096           0.0           0.0           0.0           1.0          1.0   \n",
       "HG00097           1.0           0.0           0.0           1.0          1.0   \n",
       "HG00099           0.0           0.0           0.0           1.0          0.0   \n",
       "HG00100           1.0           0.0           0.0           1.0          1.0   \n",
       "HG00101           0.0           0.0           0.0           1.0          1.0   \n",
       "\n",
       "         ...  rs4891825_GG  rs7251928_AA  rs7251928_AC  rs7251928_CC  \\\n",
       "HG00096  ...           0.0           0.0           1.0           0.0   \n",
       "HG00097  ...           0.0           1.0           0.0           0.0   \n",
       "HG00099  ...           0.0           1.0           0.0           0.0   \n",
       "HG00100  ...           0.0           1.0           0.0           0.0   \n",
       "HG00101  ...           0.0           0.0           1.0           0.0   \n",
       "\n",
       "         rs310644_CC  rs310644_CT  rs310644_TT  rs2024566_AA  rs2024566_AG  \\\n",
       "HG00096          0.0          1.0          0.0           1.0           0.0   \n",
       "HG00097          0.0          0.0          1.0           0.0           1.0   \n",
       "HG00099          0.0          0.0          1.0           0.0           1.0   \n",
       "HG00100          0.0          0.0          1.0           1.0           0.0   \n",
       "HG00101          0.0          0.0          1.0           0.0           1.0   \n",
       "\n",
       "         rs2024566_GG  \n",
       "HG00096           0.0  \n",
       "HG00097           0.0  \n",
       "HG00099           0.0  \n",
       "HG00100           0.0  \n",
       "HG00101           0.0  \n",
       "\n",
       "[5 rows x 165 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kidd_encoded.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### dimensionality reduction & training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "OVERWRITE_MODEL = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:51.376 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:51.388 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:51.401 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:51.414 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pca kidd False None\n",
      "nca kidd False None\n",
      "pca Seldin False None\n",
      "nca Seldin False None\n"
     ]
    }
   ],
   "source": [
    "# write all the super population dimred models for kidd and Seldin\n",
    "for aisnps_set, df, df_labels in zip(\n",
    "    [\"kidd\", \"Seldin\"], \n",
    "    [df_kidd_encoded, df_seldin_encoded], \n",
    "    [df_kidd[\"superpopulation\"], df_seldin[\"superpopulation\"]]\n",
    "):\n",
    "    for algorithm, labels in zip([\"pca\", \"nca\"], [None, None, None, df_labels]):\n",
    "        print(algorithm,aisnps_set,OVERWRITE_MODEL,labels)\n",
    "        df_reduced = dimensionality_reduction(df, algorithm=algorithm, aisnps_set=aisnps_set, overwrite_model=OVERWRITE_MODEL, labels=labels, population_level=\"super population\")\n",
    "        knn_model = train(df_reduced, df_labels, algorithm=algorithm, aisnps_set=aisnps_set, k=9, population_level=\"superpopulation\", overwrite_model=OVERWRITE_MODEL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:51.429 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:51.440 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n"
     ]
    }
   ],
   "source": [
    "# write all the population dimred models for kidd and Seldin\n",
    "for aisnps_set, df, df_labels in zip(\n",
    "    [\"kidd\", \"Seldin\"], \n",
    "    [df_kidd_encoded, df_seldin_encoded], \n",
    "    [df_kidd[\"population\"], df_seldin[\"population\"]]\n",
    "):\n",
    "    for algorithm, labels in zip([\"pca\"], [df_labels]):\n",
    "        df_reduced = dimensionality_reduction(df, algorithm=algorithm, aisnps_set=aisnps_set, overwrite_model=OVERWRITE_MODEL, labels=labels, population_level=\"population\")\n",
    "        knn_model = train(df_reduced, labels, algorithm=algorithm, aisnps_set=aisnps_set, k=9, population_level=\"population\", overwrite_model=OVERWRITE_MODEL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Predict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ezancestry.commands import predict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snps import SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygenomefile = \"/Users/jacksonc08/ezancestry/data/folder/HaplotypeCaller_subset.kidd.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load from DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the snps Python package will read the genome file properly \n",
    "mygenome = SNPs(mygenomefile)\n",
    "mygenomedf = mygenome.snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mygenomedf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rsid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs7554936</th>\n",
       "      <td>1</td>\n",
       "      <td>151150013</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs798443</th>\n",
       "      <td>2</td>\n",
       "      <td>7828144</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1876482</th>\n",
       "      <td>2</td>\n",
       "      <td>17181301</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs260690</th>\n",
       "      <td>2</td>\n",
       "      <td>108963282</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs6754311</th>\n",
       "      <td>2</td>\n",
       "      <td>135950412</td>\n",
       "      <td>TC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom        pos genotype\n",
       "rsid                               \n",
       "rs7554936     1  151150013       TT\n",
       "rs798443      2    7828144       GA\n",
       "rs1876482     2   17181301       GA\n",
       "rs260690      2  108963282       AA\n",
       "rs6754311     2  135950412       TC"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mygenomedf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:52.607 | INFO     | ezancestry.process:_input_to_dataframe:285 - sample has a valid genotype for 28 out of a possible 55 (50.90909090909091%)\n",
      "2022-10-25 14:12:52.622 | INFO     | ezancestry.process:encode_genotypes:137 - Successfully loaded an encoder from /Users/jacksonc08/.ezancestry/data/models/one_hot_encoder.kidd.bin\n",
      "2022-10-25 14:12:52.627 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:52.631 | INFO     | ezancestry.model:predict_ancestry:94 - Successfully loaded trained knn model: /Users/jacksonc08/.ezancestry/data/models/knn.pca.kidd.population.bin\n",
      "/Users/jacksonc08/ezancestry/ezancestry/commands.py:312: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  pop_predictions = pop_predictions.append(pop_level_predictions)\n",
      "2022-10-25 14:12:52.717 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:52.720 | INFO     | ezancestry.model:predict_ancestry:94 - Successfully loaded trained knn model: /Users/jacksonc08/.ezancestry/data/models/knn.pca.kidd.superpopulation.bin\n",
      "/Users/jacksonc08/ezancestry/ezancestry/commands.py:314: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  superpop_predictions = superpop_predictions.append(\n"
     ]
    }
   ],
   "source": [
    "# predict on the \n",
    "predictions = predict(mygenomedf, \n",
    "                    aisnps_set=\"kidd\",\n",
    "                    k=None,\n",
    "                    n_components=None,\n",
    "                    algorithm=None,\n",
    "                    write_predictions=False,\n",
    "                    models_directory=None,\n",
    "                    output_directory=None,\n",
    "                    aisnps_directory=None,\n",
    "                    thousand_genomes_directory=None,\n",
    "                    samples_directory=None\n",
    "                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>component1</th>\n",
       "      <th>component2</th>\n",
       "      <th>component3</th>\n",
       "      <th>predicted_population_population</th>\n",
       "      <th>ACB</th>\n",
       "      <th>ASW</th>\n",
       "      <th>BEB</th>\n",
       "      <th>CDX</th>\n",
       "      <th>CEU</th>\n",
       "      <th>CHB</th>\n",
       "      <th>...</th>\n",
       "      <th>TSI</th>\n",
       "      <th>YRI</th>\n",
       "      <th>predicted_population_superpopulation</th>\n",
       "      <th>AFR</th>\n",
       "      <th>AMR</th>\n",
       "      <th>EAS</th>\n",
       "      <th>EUR</th>\n",
       "      <th>SAS</th>\n",
       "      <th>population_description</th>\n",
       "      <th>superpopulation_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <td>-0.994511</td>\n",
       "      <td>-0.526379</td>\n",
       "      <td>-0.05067</td>\n",
       "      <td>ITU</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SAS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.103656</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.896344</td>\n",
       "      <td>Indian Telugu in the UK</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        component1  component2  component3 predicted_population_population  \\\n",
       "sample   -0.994511   -0.526379    -0.05067                             ITU   \n",
       "\n",
       "        ACB  ASW  BEB  CDX  CEU  CHB  ...  TSI  YRI  \\\n",
       "sample  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   \n",
       "\n",
       "        predicted_population_superpopulation  AFR       AMR  EAS  EUR  \\\n",
       "sample                                   SAS  0.0  0.103656  0.0  0.0   \n",
       "\n",
       "             SAS   population_description  superpopulation_name  \n",
       "sample  0.896344  Indian Telugu in the UK  South Asian Ancestry  \n",
       "\n",
       "[1 rows x 38 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## or load directly from a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-25 14:12:52.782 | INFO     | ezancestry.process:_input_to_dataframe:285 - HaplotypeCaller_subset.kidd.vcf.gz has a valid genotype for 28 out of a possible 55 (50.90909090909091%)\n",
      "2022-10-25 14:12:52.796 | INFO     | ezancestry.process:encode_genotypes:137 - Successfully loaded an encoder from /Users/jacksonc08/.ezancestry/data/models/one_hot_encoder.kidd.bin\n",
      "2022-10-25 14:12:52.800 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:52.803 | INFO     | ezancestry.model:predict_ancestry:94 - Successfully loaded trained knn model: /Users/jacksonc08/.ezancestry/data/models/knn.nca.kidd.population.bin\n",
      "/Users/jacksonc08/ezancestry/ezancestry/commands.py:312: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  pop_predictions = pop_predictions.append(pop_level_predictions)\n",
      "2022-10-25 14:12:52.822 | INFO     | ezancestry.dimred:dimensionality_reduction:126 - Successfully loaded a dimensionality reduction model\n",
      "2022-10-25 14:12:52.824 | INFO     | ezancestry.model:predict_ancestry:94 - Successfully loaded trained knn model: /Users/jacksonc08/.ezancestry/data/models/knn.nca.kidd.superpopulation.bin\n",
      "/Users/jacksonc08/ezancestry/ezancestry/commands.py:314: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  superpop_predictions = superpop_predictions.append(\n"
     ]
    }
   ],
   "source": [
    "predictions = predict(mygenomefile, \n",
    "                    aisnps_set=\"kidd\",\n",
    "                    k=None,\n",
    "                    n_components=None,\n",
    "                    algorithm=\"nca\",\n",
    "                    write_predictions=False,\n",
    "                    models_directory=None,\n",
    "                    output_directory=None,\n",
    "                    aisnps_directory=None,\n",
    "                    thousand_genomes_directory=None,\n",
    "                    samples_directory=None\n",
    "                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>component1</th>\n",
       "      <th>component2</th>\n",
       "      <th>component3</th>\n",
       "      <th>predicted_population_population</th>\n",
       "      <th>ACB</th>\n",
       "      <th>ASW</th>\n",
       "      <th>BEB</th>\n",
       "      <th>CDX</th>\n",
       "      <th>CEU</th>\n",
       "      <th>CHB</th>\n",
       "      <th>...</th>\n",
       "      <th>TSI</th>\n",
       "      <th>YRI</th>\n",
       "      <th>predicted_population_superpopulation</th>\n",
       "      <th>AFR</th>\n",
       "      <th>AMR</th>\n",
       "      <th>EAS</th>\n",
       "      <th>EUR</th>\n",
       "      <th>SAS</th>\n",
       "      <th>population_description</th>\n",
       "      <th>superpopulation_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HaplotypeCaller_subset.kidd.vcf.gz</th>\n",
       "      <td>19.780208</td>\n",
       "      <td>15.782891</td>\n",
       "      <td>18.565223</td>\n",
       "      <td>MXL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AMR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>American Ancestry</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    component1  component2  component3  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz   19.780208   15.782891   18.565223   \n",
       "\n",
       "                                   predicted_population_population  ACB  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz                             MXL  0.0   \n",
       "\n",
       "                                         ASW  BEB  CDX  CEU  CHB  ...  TSI  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz  0.083226  0.0  0.0  0.0  0.0  ...  0.0   \n",
       "\n",
       "                                    YRI  predicted_population_superpopulation  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz  0.0                                   AMR   \n",
       "\n",
       "                                    AFR  AMR  EAS  EUR  SAS  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz  0.0  1.0  0.0  0.0  0.0   \n",
       "\n",
       "                                                         population_description  \\\n",
       "HaplotypeCaller_subset.kidd.vcf.gz  Mexican Ancestry in Los Angeles, California   \n",
       "\n",
       "                                    superpopulation_name  \n",
       "HaplotypeCaller_subset.kidd.vcf.gz     American Ancestry  \n",
       "\n",
       "[1 rows x 38 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = predictions.to_csv(\"/Users/jacksonc08/.ezancestry/predictions.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "mypredictions = SNPs(predictions)\n",
    "mypredictionsf = mypredictions.snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ezancestry.commands import plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "expected str, bytes or os.PathLike object, not OptionInfo",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [34]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m plots \u001b[38;5;241m=\u001b[39m \u001b[43mplot\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmypredictionsf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;43;03m#             population_level=\"superpopulation\",\u001b[39;49;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;43;03m#             output_directory=None\u001b[39;49;00m\n\u001b[1;32m      4\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/ezancestry/ezancestry/commands.py:374\u001b[0m, in \u001b[0;36mplot\u001b[0;34m(predictions_file, population_level, output_directory)\u001b[0m\n\u001b[1;32m    372\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m output_directory \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    373\u001b[0m     output_directory \u001b[38;5;241m=\u001b[39m Path\u001b[38;5;241m.\u001b[39mcwd()\n\u001b[0;32m--> 374\u001b[0m output_directory \u001b[38;5;241m=\u001b[39m \u001b[43mPath\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput_directory\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    375\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m population_level \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    376\u001b[0m     population_level \u001b[38;5;241m=\u001b[39m _population_level\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/pathlib.py:1082\u001b[0m, in \u001b[0;36mPath.__new__\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1080\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m Path:\n\u001b[1;32m   1081\u001b[0m     \u001b[38;5;28mcls\u001b[39m \u001b[38;5;241m=\u001b[39m WindowsPath \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnt\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m PosixPath\n\u001b[0;32m-> 1082\u001b[0m \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_from_parts\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minit\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   1083\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_flavour\u001b[38;5;241m.\u001b[39mis_supported:\n\u001b[1;32m   1084\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcannot instantiate \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m on your system\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1085\u001b[0m                               \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m,))\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/pathlib.py:707\u001b[0m, in \u001b[0;36mPurePath._from_parts\u001b[0;34m(cls, args, init)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_from_parts\u001b[39m(\u001b[38;5;28mcls\u001b[39m, args, init\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;66;03m# We need to call _parse_args on the instance, so as to get the\u001b[39;00m\n\u001b[1;32m    705\u001b[0m     \u001b[38;5;66;03m# right flavour.\u001b[39;00m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__new__\u001b[39m(\u001b[38;5;28mcls\u001b[39m)\n\u001b[0;32m--> 707\u001b[0m     drv, root, parts \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parse_args\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    708\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_drv \u001b[38;5;241m=\u001b[39m drv\n\u001b[1;32m    709\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_root \u001b[38;5;241m=\u001b[39m root\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/pathlib.py:691\u001b[0m, in \u001b[0;36mPurePath._parse_args\u001b[0;34m(cls, args)\u001b[0m\n\u001b[1;32m    689\u001b[0m     parts \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m a\u001b[38;5;241m.\u001b[39m_parts\n\u001b[1;32m    690\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 691\u001b[0m     a \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfspath\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    692\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(a, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    693\u001b[0m         \u001b[38;5;66;03m# Force-cast str subclasses to str (issue #21127)\u001b[39;00m\n\u001b[1;32m    694\u001b[0m         parts\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28mstr\u001b[39m(a))\n",
      "\u001b[0;31mTypeError\u001b[0m: expected str, bytes or os.PathLike object, not OptionInfo"
     ]
    }
   ],
   "source": [
    "plots = plot(mypredictionsf, \n",
    "#             population_level=\"superpopulation\",\n",
    "#             output_directory=None\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.system(\"ezancestry plot predictions.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
